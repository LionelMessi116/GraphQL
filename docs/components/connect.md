### About

Connect is a ReactJS component that is used to execute queries and mutations and render child components with the results, using a render prop.

### Props

| Name      | Type                                                         | Default   | Description                                                        |
| --------- | ------------------------------------------------------------ | --------- | ------------------------------------------------------------------ |
| query     | [Query](../types/mutation-query.md)?                         | undefined | The query you want connected to your component                     |
| mutations | Object { [string]: [Mutation](../types/mutation-query.md) }? | undefined | The mutation/mutations you want connected to your component        |
| children  | ([ChildArgs\<T\>](../types/child-args.md)) => ReactNode      | undefined | A child function to accept the connected state and render elements |

### Example

```jsx
import { Connect, createQuery, createMutation } from 'urql';

// ...
<Connect
  query={createQuery(GetTodos)}
  mutations={{
    addTodo: createMutation(AddTodo)
  }}
  children={({ fetching, data }) => {
    return fetching ? <Loading/> : <List data={data.todos}>
  }}
/>
```
