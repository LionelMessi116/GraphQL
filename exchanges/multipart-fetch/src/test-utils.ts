import {
  GraphQLRequest,
  OperationContext,
  Operation,
  makeOperation,
} from '@urql/core';
import gql from 'graphql-tag';

const context: OperationContext = {
  fetchOptions: {
    method: 'POST',
  },
  requestPolicy: 'cache-first',
  url: 'http://localhost:3000/graphql',
};

const queryGql: GraphQLRequest = {
  key: 2,
  query: gql`
    query getUser($name: String) {
      user(name: $name) {
        id
        firstName
        lastName
      }
    }
  `,
  variables: {
    name: 'Clara',
  },
};

const obj = { hello: 'world' };
const file = new File(
  [new Blob([JSON.stringify(obj, null, 2)], { type: 'application/json' })],
  'index.ts'
);
const files = [file, file];

const upload: GraphQLRequest = {
  key: 3,
  query: gql`
    mutation uploadProfilePicture($picture: File) {
      uploadProfilePicture(picture: $picture) {
        location
      }
    }
  `,
  variables: {
    picture: file,
  },
};

const uploads: GraphQLRequest = {
  key: 3,
  query: gql`
    mutation uploadProfilePictures($pictures: [File]) {
      uploadProfilePicture(pictures: $pictures) {
        location
      }
    }
  `,
  variables: {
    picture: files,
  },
};

export const uploadOperation: Operation = makeOperation(
  'mutation',
  upload,
  context
);

export const multipleUploadOperation: Operation = makeOperation(
  'mutation',
  uploads,
  context
);

export const queryOperation: Operation = makeOperation(
  'query',
  queryGql,
  context
);
