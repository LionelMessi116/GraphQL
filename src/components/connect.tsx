import React, { ReactNode } from 'react';
import { ChildArgs, Client, Mutation, Query } from '../types';
import { UrqlClient } from './client';
import { ContextConsumer } from './context';

/** Props for the [Connect]{@link Connect} component. */
export interface ConnectProps<T> {
  /** A function which receives values from the URQL client. */
  children: (props: ChildArgs<T>) => ReactNode;
  /** The GraphQL query to fetch */
  query?: Query;
  /** A collection of GrahpQL mutation queries */
  mutations?: { [type in keyof T]: Mutation };
}

/** Component for connecting to the urql client for executing queries, mutations and returning the result to child components. */
export const Connect = function<T>(props: ConnectProps<T>) {
  return (
    <ContextConsumer>
      {(client: Client) => (
        <UrqlClient
          client={client}
          children={props.children}
          query={props.query}
          mutations={props.mutations}
        />
      )}
    </ContextConsumer>
  );
};
