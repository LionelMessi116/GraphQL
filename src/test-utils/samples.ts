import {
  ExchangeResult,
  ExecutionResult,
  Mutation,
  Operation,
  Query,
  Subscription,
} from '../types';

const context = {
  fetchOptions: {
    method: 'POST',
  },
  url: 'http://localhost:3000/graphql',
};

export const queryGql: Query = {
  query: `query getUser($name: String){
    user(name: $name) {
      id
      firstName
      lastName
    }
  }`,
  variables: {
    name: 'Clara',
  },
};

export const mutationGql: Mutation = {
  query: `mutation AddUser($name: String){
    addUser(name: $name) {
      name
    }
  }`,
  variables: {
    name: 'Clara',
  },
};

export const subscriptionGql: Subscription = {
  query: `subscription subscribeToUser($user: String){
    user(user: $user) {
      name
    }
  }
  `,
  variables: {
    user: 'colin',
  },
};

export const queryOperation: Operation = {
  key: '2',
  operationName: 'query',
  context,
  ...queryGql,
};

export const mutationOperation: Operation = {
  key: JSON.stringify(mutationGql),
  operationName: 'mutation',
  context,
  ...mutationGql,
};

export const subscriptionOperation: Operation = {
  key: JSON.stringify(subscriptionGql),
  operationName: 'subscription',
  context,
  ...subscriptionGql,
};

export const queryResponse: ExchangeResult = {
  operation: queryOperation,
  data: {
    user: {
      name: 'Clive',
    },
  },
};

export const mutationResponse: ExchangeResult = {
  operation: mutationOperation,
  data: {},
};

export const subscriptionResult: ExecutionResult = {
  data: {}
};
