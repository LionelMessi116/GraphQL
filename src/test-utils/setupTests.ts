import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

// Setup enzyme adapter
configure({ adapter: new Adapter() });

// Setup globals
const g = global as any;

g.AbortController = undefined;
g.fetch = jest.fn();

process.on('unhandledRejection', error => {
  throw error;
});
