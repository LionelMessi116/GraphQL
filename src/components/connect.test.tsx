jest.mock('./client');
import { shallow, mount } from 'enzyme';
import React from 'react';
import { Connect } from './connect';
import { queryGql, mutationGql } from '../test-utils';

const children = jest.fn().mockReturnValue(() => <h1>Child component</h1>);
const query = queryGql;
const mutations = {
  addUser: mutationGql,
};

beforeEach(() => {
  children.mockClear();
});

it('passes snapshot', () => {
  const wrapper = shallow(
    <Connect children={children} mutations={mutations} query={query} />
  );
  expect(wrapper).toMatchSnapshot();
});

it('provides context consumer', () => {
  const wrapper = shallow(<Connect children={children} />);
  expect(wrapper.find('ContextConsumer').length).toBe(1);
});

describe('UrqlClient', () => {
  it('is mounted', () => {
    const wrapper = mount(<Connect children={children} />);
    expect(wrapper.find('UrqlClient').length).toBe(1);
  });

  it('is passed children', () => {
    const wrapper = mount(<Connect children={children} />);
    expect(wrapper.find('UrqlClient').props()).toHaveProperty(
      'children',
      children
    );
  });

  it('is passed query', () => {
    const wrapper = mount(<Connect children={children} query={query} />);
    expect(wrapper.find('UrqlClient').props()).toHaveProperty('query', query);
  });

  it('is passed mutation', () => {
    const wrapper = mount(
      <Connect children={children} mutations={mutations} />
    );
    expect(wrapper.find('UrqlClient').props()).toHaveProperty(
      'mutations',
      mutations
    );
  });

  it('is passed client', () => {
    const wrapper = mount(<Connect children={children} />);
    expect(wrapper.find('UrqlClient').props()).toHaveProperty('client', {});
  });
});
