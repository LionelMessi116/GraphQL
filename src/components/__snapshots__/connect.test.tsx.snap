// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`passes snapshot 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;
